### PR TITLE
[SRVKS-1189] Generate istio net policies if we have to

### PR DESCRIPTION
**Pull request by skonto**
Fixes JIRA #SRVKS-1189

## Proposed Changes

-  By default when OSSM is installed we get (absence implies true):
```
  profiles:
  - default
  security:
    manageNetworkPolicy: true
```
In this case two default net OSSM policies are created [nets.txt](https://github.com/openshift-knative/serverless-operator/files/14097615/nets.txt), one for allowing access from openshift routes and another that allows access to pods with the following label set:
```
                "podSelector": {
                    "matchLabels": {
                        "maistra.io/expose-route": "true"
                    }
                },
````

In order to integrate with those default net policies we generate three extra Serverless network policies when istio is enabled:

   a) webhook
   b) net-istio-webhook
   c) allow-from-openshift-monitoring-ns

The first two are meant to allow access from the  Api server to the webhooks and the last one for Prometheus to be able to scrape the Serving pods.

- We this patch we provide the option to disable the generation of our net policies when user sets: `manageNetworkPolicy: false` . With no net policies traffic works as expected. Also enabling monitoring by setting in Serving CR (injects kube-rbac-proxy)
```
  config:
    observability:
      metrics.backend-destination: "prometheus"
```
and by setting in smcp:
```
  proxy:
    networking:
      trafficControl:
        inbound:
          excludedPorts:
          - 8444 # metrics
          - 8022 # serving: wait-for-drain k8s pre-stop hook
```
allows prometheus to scrape targets without any net policies.

image
In other words having no net policies relaxes restrictions.
- TODO: test midstream serving against removing the net policies (I think we don't have tests for istio there).

### CODE REVIEW

**Comment by skonto** [comment]
```
 TestControllerHA
```
I think we have a ticket for that one since it is flaky for ages.

**Comment by skonto** [comment]
@ReToCode ready again, had to rebase.

**Comment by skonto** [comment]
@ReToCode ready for stamping.

File: openshift-knative-operator/pkg/serving/ingress.go [hunk 1]
```diff
@@ -51,3 +62,57 @@ func defaultIngressClass(ks *operatorv1beta1.KnativeServing) string {
...
 	return kourierIngressClassName
 }
+
+func generateDefaultIstioNetworkPoliciesIfRequired(ks base.KComponent) ([]mf.Manifest, error) {
+	if !ks.(*operatorv1beta1.KnativeServing).Spec.Ingress.Istio.Enabled {
+		return nil, nil
+	}
+
+	if v, ok := ks.GetAnnotations()[disableGeneratingIstioNetPoliciesAnnotation]; ok {
+		b, _ := strconv.ParseBool(v)
```
Comment (ReToCode) [inline openshift-knative-operator/pkg/serving/ingress.go:72]: maybe name this variable `disableGeneration` just to make it easier to read

File: openshift-knative-operator/pkg/serving/ingress.go [hunk 2]
```diff
@@ -51,3 +62,57 @@ func defaultIngressClass(ks *operatorv1beta1.KnativeServing) string {
...
+		nwp := networkingv1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: ks.GetNamespace(),
+				Labels: map[string]string{
+					"networking.knative.dev/ingress-provider": "istio",
+				},
+			},
+		}
+		if name != "allow-from-openshift-monitoring-ns" {
```
Comment (ReToCode) [inline openshift-knative-operator/pkg/serving/ingress.go:89]: maybe use `switch` to make it more clear which cases are the same?

**Review by ReToCode — COMMENTED** [review]
See inline

File: openshift-knative-operator/pkg/serving/ingress.go (hunk 2 above)
Comment (skonto) [reply openshift-knative-operator/pkg/serving/ingress.go:89]: Can do that

**Review by ReToCode — APPROVED** [review]
/lgtm
/approve

**Comment by skonto** [comment]
@ReToCode should we merge?
Upgrade tests fail due to: `BrokerBackedByChannelContinualTest` and Kafka one never passed so far.
@pierDipi should we wait for any changes?

**Comment by ReToCode** [comment]
I can't really tell. Deferring to Eventing team :)

**Comment by pierDipi** [comment]
Yes, we can merge, we're working on the Kafka one in the matching midstream branch.

The upgrade one is new

/lgtm
/unhold
/approve
